### PR TITLE
Fix thread destructors

### PR DESCRIPTION
**Pull request by zzmp**
- Properly clean up threads by terminating them from the virtual dtors, not base dtor
- Avoid crash from moving finished GL context back to main thread

### CODE REVIEW

**Comment by zzmp** [comment]
See: https://app.asana.com/0/74051501495175/110582563415251.

This does not occur in Release builds, because in release `main.cpp` exits using `_exit`.
This allows a clean exit on Debug builds, or builds with `_exit` commented out, by making sure the threads that are destructed on shutdown do not crash.

#### Testing

General smoke test.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4cedf9853320a8633d0ae94784c94570e981b344 available at: https://gist.github.com/33ca7e5bea7b05f5b8ff26e7c3bf8371<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e2b87acb83860e92ba21d87d1e9e2faa0f8dc111 available at: https://gist.github.com/33ca7e5bea7b05f5b8ff26e7c3bf8371<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.dmg

**Comment by ZappoMan** [comment]
This change is going to require an extensive test. Not just a smoke test.

**Comment by hifi-gustavo** [comment]
Build console output for commit: dc5df0dc3b0157561c304640fca89377436783c4 available at: https://gist.github.com/33ca7e5bea7b05f5b8ff26e7c3bf8371<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a973698b8d7bee7b23581bb8ace7500ffa2fe9e1 available at: https://gist.github.com/33ca7e5bea7b05f5b8ff26e7c3bf8371<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.dmg

**Comment by zzmp** [comment]
Still crashing on UNIX, testing with Seth.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6dc2c2aa8d9b7597725e7bf207dafcdd436601a3 available at: https://gist.github.com/33ca7e5bea7b05f5b8ff26e7c3bf8371<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.dmg

File: libraries/gpu/src/gpu/GLBackendTextureTransfer.h [hunk 1]
```diff
@@ -23,12 +23,14 @@ struct TextureTransferPackage {
...
     GLTextureTransferHelper();
+    ~GLTextureTransferHelper() { cleanup(); }
+
     void transferTexture(const gpu::TexturePointer& texturePointer);
     void postTransfer(const gpu::TexturePointer& texturePointer);
 
 protected:
     void setup() override;
-    void shutdown() override;
+    void shutdown() override { _canvas.reset(); }
```
Comment (jherico) [inline libraries/gpu/src/gpu/GLBackendTextureTransfer.h]: This will cause the GL canvas to be destroyed while it is still active.

File: libraries/gpu/src/gpu/GLBackendTextureTransfer.cpp [hunk 2]
```diff
@@ -77,12 +77,6 @@ void GLTextureTransferHelper::setup() {
 #endif
 }
 
-void GLTextureTransferHelper::shutdown() {
-    _canvas->doneCurrent();
```
Comment (jherico) [inline libraries/gpu/src/gpu/GLBackendTextureTransfer.cpp]: why did you remove this logic.  the GLTextureTransferHelper is created on the main thread and will be destroyed on the main thread.  The context needs to be deactivated and moved back to the main thread prior to destruction.

File: libraries/shared/src/GenericThread.cpp [hunk 3]
```diff
@@ -21,7 +21,7 @@ GenericThread::GenericThread() :
     
 }
 
-GenericThread::~GenericThread() {
```
Comment (jherico) [inline libraries/shared/src/GenericThread.cpp]: You should add a member variable that it set to true when cleanup is called, and check that variable in the GenericThread destructor to ensure it was called.

File: libraries/gpu/src/gpu/GLBackendTextureTransfer.cpp (hunk 2 above)
Comment (zzmp) [inline libraries/gpu/src/gpu/GLBackendTextureTransfer.cpp]: Why does it need to be destroyed on the thread on which it was created? It calls doneCurrent in the dtor, that's why I'd refactored like this, I'd assumed it was enough.

File: libraries/gpu/src/gpu/GLBackendTextureTransfer.h (hunk 1 above)
Comment (zzmp) [inline libraries/gpu/src/gpu/GLBackendTextureTransfer.h]: Why is it still active? How is it deactivated?

File: libraries/shared/src/GenericThread.cpp (hunk 3 above)
Comment (zzmp) [inline libraries/shared/src/GenericThread.cpp]: If it's not multithreaded is this necessary? Why don't you think it's enough to make sure the thread's no longer running?

Comment (zzmp) [inline libraries/shared/src/GenericThread.cpp]: I took it out because it was crashing, but I think it may have been accessing qApp->thread after qApp was destroyed. On testing, the access violation is actually at qApp->thread().

**Comment by hifi-gustavo** [comment]
Build console output for commit: 541628e101764c88865db18dd882c0909816e5bc available at: https://gist.github.com/33ca7e5bea7b05f5b8ff26e7c3bf8371<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: df63d051e1dbe27b4d0ff939b55b31967dae79df available at: https://gist.github.com/33ca7e5bea7b05f5b8ff26e7c3bf8371<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7629.dmg

**Comment by zzmp** [comment]
Too big of a change to a core internal. Closing.
